libs/core/lci_base/src/lci_environment.cpp

```diff
@@ -214,17 +214,22 @@ namespace hpx { namespace util {
             LCI_initialized(&lci_init);
             if (lci_init)
             {
-                HPX_ASSERT(prg_thread_flag.load() == true);
-                HPX_ASSERT(prg_thread_p != nullptr);
-                prg_thread_flag = false;
-                prg_thread_p->join();
-                delete prg_thread_p;
-
+                join_prg_thread_if_running();
                 LCI_finalize();
             }
         }
     }
 
+    void lci_environment::join_prg_thread_if_running()
+    {
+        if (prg_thread_flag.load())
```

Comment:
Should this be a `compare_exchange` instead? The way it's coded now the atomic doesn't protect from races.

Reply:
For this specific case, only the thread belonging to the `lci-progress-pool` will call this function, so it will not be called by multiple threads at the same time. However, I admit the code here can be tricky. It would be much better if there is a parcelport function that is called after all the early parcels are sent/received and before the scheduling loop of worker threads starts. Is it possible to have this function?

Reply:
I used a lock to make it thread-safe for now, but a cleaner way would be to provide a method in the parcelport interface to join the progress thread after all the early parcels are sent/received and before any worker threads can call the do_background_work function.

Reply:
> For this specific case, only the thread belonging to the lci-progress-pool will call this function, so it will not be called by multiple threads at the same time.

Why did you use an atomic, then?

> It would be much better if there is a parcelport function that is called after all the early parcels are sent/received and before the scheduling loop of worker threads starts. Is it possible to have this function?

Unfortunately, that cannot be done as the early parcels are being sent only after the schedulers have been brought up. The handling of the early parcels requires for the threading to be functional. I can however notify the parcelports that the early parcel handling is done and normal operation is about to start. Is that what you need?

Reply:
> Why did you use an atomic, then?

This variable is written by the thread in lci-progress-pool and read by the thread pointed by prg_thread_p, so I make it an atomic.

> Unfortunately, that cannot be done as the early parcels are being sent only after the schedulers have been brought up. The handling of the early parcels requires for the threading to be functional.

By "the schedulers have been brought up", do you mean the worker threads can already call the do_background_work function? If that is the case, why the thread in lci-progress-pool is not functioning (no LCI_progress is called) when early parcels are sent?

Reply:
> By "the schedulers have been brought up", do you mean the worker threads can already call the do_background_work function? 

I think so, yes.

> If that is the case, why the thread in lci-progress-pool is not functioning (no LCI_progress is called) when early parcels are sent?

I'm not sure, will investigate.

Reply:
@JiakunYan you were right. The pools start running only after early parcel handling (that's also the reason why you need your own thread for those). I will add the notification function as you requested.

Reply:
I figured out early parcels are sent in `big_boot_barrier`. Could you point me to the function where schedulers are brought up?

Reply:
> I will add the notification function as you requested.

That's great! Thank you!